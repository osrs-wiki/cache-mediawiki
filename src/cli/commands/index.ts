export { default as combatAchievements } from "./combatAchievements";
export { default as clues } from "./clues";
export { default as differences } from "./differences";
