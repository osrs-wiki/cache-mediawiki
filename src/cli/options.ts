import { Option } from "commander";

const options = [
  new Option(
    "-n, --newCache <newCache>",
    "Path to the new cache."
  ).makeOptionMandatory(),
  new Option("-o, --oldCache <oldCache>", "Path to the old cache."),
  new Option("--cacheSource <source>", "Source of the cache.")
    .choices(["github", "local"])
    .default("github"),
  new Option("--cacheType <type>", "Type of the cache.")
    .choices(["disk", "flat"])
    .default("flat"),
  new Option("--renders <renders>", "Include renders in the output.").default(
    false
  ),
  new Option(
    "--infoboxes <infoboxes>",
    "Include pages with infoboxes in the output."
  ).default(false),
  new Option("--update <update>", "The news post update title."),
  new Option("--updateDate <updateDate>", "The news post update date."),
  new Option(
    "--examines <examines>",
    "Include examines in the output."
  ).default(false),
  new Option(
    "--examinesVersion <examinesVersion>",
    "The version of the examines."
  ).default("latest"),
];

export default options;
