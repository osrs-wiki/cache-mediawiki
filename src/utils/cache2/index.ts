export * from "./LazyPromise";
export * from "./types";
