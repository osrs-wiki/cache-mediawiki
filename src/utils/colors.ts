export const jagexHSLtoHex = (value: number) => {
  const hue = (value >> 10) & 0x3f;
  const sat = (value >> 7) & 0x07;
  const lum = value & 0x7f;

  console.log(`H: ${hue}, S: ${sat}, L: ${lum}`);
  return hslToHex(hue, sat, lum);
};

export const hslToHex = (h: number, s: number, l: number) => {
  l /= 100;
  const a = (s * Math.min(l, 1 - l)) / 100;
  const f = (n: number) => {
    const k = (n + h / 30) % 12;
    const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);
    return Math.round(255 * color)
      .toString(16)
      .padStart(2, "0"); // convert to Hex and prefix "0" if needed
  };
  return `#${f(0)}${f(8)}${f(4)}`;
};
