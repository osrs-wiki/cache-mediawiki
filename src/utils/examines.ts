export const getExamines = async (
  name: string,
  version = "master"
): Promise<{ [key: string]: string }> => {
  const examines: { [key: string]: string } = {};
  const req = await fetch(
    `https://raw.githubusercontent.com/Joshua-F/osrs-examines/${version}/${name}.csv`
  );
  if (!req.ok) {
    return;
  }
  const fetchedExamines = await req.text();
  fetchedExamines.split("\n").forEach((line) => {
    const id = line.split(",")[0];
    const examine = line.substring(line.indexOf(",") + 1).replaceAll('"', "");
    if (id) {
      examines[id] = examine;
    }
  });
  return examines;
};
