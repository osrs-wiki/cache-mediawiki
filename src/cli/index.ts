import cli from "./program";

export default cli;
