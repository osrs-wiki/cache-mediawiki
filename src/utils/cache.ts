import { DiskCacheProvider, FlatCacheProvider } from "@abextm/cache2";
import { readFile } from "fs/promises";

export type CacheSource = "github" | "local";
export type CacheFileType = "flat" | "disk";

export const getCacheProviderLocal = async (
  version: string,
  type: CacheFileType
) => {
  if (type === "disk") {
    return getCacheProviderDiskLocal(version);
  } else if (type === "flat") {
    return getCacheProviderFlatLocal(version);
  }
};

export const getCacheProviderDiskLocal = async (version: string) => {
  return new DiskCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/diskcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderFlatLocal = async (version: string) => {
  return new FlatCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/flatcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderGithub = async (version = "master") => {
  return new FlatCacheProvider({
    async getFile(name) {
      const req = await fetch(
        `https://raw.githubusercontent.com/abextm/osrs-cache/${version}/${name}`
      );
      if (!req.ok) {
        return;
      }
      const ab = await req.arrayBuffer();
      return new Uint8Array(ab);
    },
  });
};
