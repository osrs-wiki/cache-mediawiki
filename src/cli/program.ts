import { program } from "commander";

import { combatAchievements, clues, differences, pages } from "./commands";
import options from "./options";
import packageJson from "../../package.json";

import Context from "@/context";
import { getExamines } from "@/utils/examines";
import { getLatestNewsTitle } from "@/utils/news";

const commands = [combatAchievements, clues, differences, pages];

program
  .name("Cache to MediaWiki tools")
  .description(
    "A set of scripts for fetching content from the OSRS cache and transforming it into MediaWiki format."
  )
  .version(packageJson.version)
  .hook("preAction", async (_program, command) => {
    const options = command.optsWithGlobals();
    Context.beta = options.beta;
    Context.pages = options.pages;
    Context.update = options.update;
    Context.updateDate = options.updateDate;

    const renders = options.renders;
    Context.renders =
      renders === "true"
        ? "renders"
        : renders === "false"
        ? undefined
        : renders;

    if (Context.update === "auto") {
      const latestUpdate = await getLatestNewsTitle();
      if (latestUpdate) {
        Context.update = latestUpdate.title;
        Context.updateDate = latestUpdate.date;
      }
    }

    if (options.examines) {
      const version = options.examinesVersion;
      Context.examines = {
        npcs: {},
        scenery: {},
      };

      const npcExamines = await getExamines("npcs", version);
      Object.keys(npcExamines).forEach((key) => {
        Context.examines.npcs[key] = npcExamines[key];
      });

      const objectExamines = await getExamines("locs", version);
      Object.keys(objectExamines).forEach((key) => {
        Context.examines.scenery[key] = objectExamines[key];
      });
    }
  });

commands.forEach((command) => {
  options.forEach((option) => {
    command.addOption(option);
  });
  program.addCommand(command);
});

export default program;
