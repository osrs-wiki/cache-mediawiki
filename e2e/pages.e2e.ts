import { existsSync } from "fs";

import { runCLICommand, cleanupOutput } from "./utils/cli";

describe("Pages Command E2E Tests", () => {
  afterEach(async () => {
    await cleanupOutput();
  });

  describe("Item Pages", () => {
    it("should generate item page with expected output file structure", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "item",
        ],
        expectedOutputFile: "./out/pages/item/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(existsSync("./out/pages/item/1.txt")).toBe(true);
      expect(existsSync("./out/pages/item/named/Toolkit.txt")).toBe(true);
    });

    it("should generate item page content that matches snapshot", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "item",
        ],
        expectedOutputFile: "./out/pages/item/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(result.outputContent).toMatchSnapshot("item-1-output.txt");
    });

    it("should generate different item page for different ID", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "2",
          "--type",
          "item",
        ],
        expectedOutputFile: "./out/pages/item/2.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(result.outputContent).toMatchSnapshot("item-2-output.txt");
    });
  });

  describe("NPC Pages", () => {
    it("should generate npc page with expected output file structure", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "npc",
        ],
        expectedOutputFile: "./out/pages/npc/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(existsSync("./out/pages/npc/1.txt")).toBe(true);
    });

    it("should generate npc page content that matches snapshot", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "npc",
        ],
        expectedOutputFile: "./out/pages/npc/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(result.outputContent).toMatchSnapshot("npc-1-output.txt");
    });
  });

  describe("Area Pages", () => {
    it("should generate area page with expected output file structure", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "area",
        ],
        expectedOutputFile: "./out/pages/area/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(existsSync("./out/pages/area/1.txt")).toBe(true);
    });

    it("should generate area page content that matches snapshot", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "area",
        ],
        expectedOutputFile: "./out/pages/area/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(result.outputContent).toMatchSnapshot("area-1-output.txt");
    });
  });

  describe("Scenery Pages", () => {
    it("should generate scenery page with expected output file structure", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "scenery",
        ],
        expectedOutputFile: "./out/pages/scenery/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(existsSync("./out/pages/scenery/1.txt")).toBe(true);
    });

    it("should generate scenery page content that matches snapshot", async () => {
      const result = await runCLICommand({
        command: "npm run start",
        args: [
          "pages",
          "--",
          "--newCache",
          "test-cache",
          "--id",
          "1",
          "--type",
          "scenery",
        ],
        expectedOutputFile: "./out/pages/scenery/1.txt",
      });

      expect(result.exitCode).toBe(0);
      expect(result.outputContent).toMatchSnapshot("scenery-1-output.txt");
    });
  });
});
