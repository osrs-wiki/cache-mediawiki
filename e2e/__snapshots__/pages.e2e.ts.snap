// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Command E2E Tests Area Pages should generate area page content that matches snapshot: area-1-output.txt 1`] = `
"{{New Content}}
{{Infobox Location
|image = [[File:undefined.png|300px]]
|members = Yes
|location = 
|wilderness = 
|teleport = 
|music = 
|map = 
|type = 
}}
[[File:undefined.png|300px|left]]
 is an area."
`;

exports[`Pages Command E2E Tests Item Pages should generate different item page for different ID: item-2-output.txt 1`] = `
"{{New Content}}
{{Infobox Item
|name = Cannonball
|image = [[File:Cannonball.png]]
|members = Yes
|quest = No
|exchange = Yes
|tradeable = Yes
|bankable = Yes
|placeholder = Yes
|equipable = No
|stackable = Yes
|noteable = No
|options = Drop
|examine = Ammo for the Dwarf Cannon.
|value = 5
|weight = 3.628
|id = 2
}}
[[File:Cannonball detail.png|130px|left]]
'''Cannonball''' is an item."
`;

exports[`Pages Command E2E Tests Item Pages should generate item page content that matches snapshot: item-1-output.txt 1`] = `
"{{New Content}}
{{Infobox Item
|name = Toolkit
|image = [[File:Toolkit.png]]
|members = Yes
|quest = No
|exchange = No
|tradeable = No
|bankable = Yes
|placeholder = Yes
|equipable = No
|stackable = No
|noteable = No
|options = Destroy
|examine = Good for repairing broken cannons.
|value = 1
|weight = 0.453
|id = 1
}}
[[File:Toolkit detail.png|130px|left]]
'''Toolkit''' is an item."
`;

exports[`Pages Command E2E Tests NPC Pages should generate npc page content that matches snapshot: npc-1-output.txt 1`] = `
"{{New Content}}
{{Infobox Monster
|name = Molanisk
|image = [[File:Molanisk.png|120px]]
|members = Yes
|combat = 51
|size = 1
|examine = 
|hitpoints = 52
|att = 40
|def = 50
|mage = 0
|dstab = 45
|dslash = 45
|dcrush = 35
|dmagic = 30
|id = 1
}}
'''Molanisk''' is an NPC."
`;

exports[`Pages Command E2E Tests Scenery Pages should generate scenery page content that matches snapshot: scenery-1-output.txt 1`] = `
"{{New Content}}
{{Infobox Scenery
|name = Crate
|image = [[File:Crate.png]]
|members = Yes
|quest = No
|options = Search
|examine = 
|map = No
|id = 1
}}
'''Crate''' is a scenery object."
`;
